src/cryptonote_core/cryptonote_core.cpp

```diff
@@ -1699,20 +1699,88 @@ namespace cryptonote
     return m_mempool.check_for_key_images(key_im, spent);
   }
   //-----------------------------------------------------------------------------------------------
-  std::tuple<uint64_t, uint64_t, uint64_t> core::get_coinbase_tx_sum(const uint64_t start_offset, const size_t count)
+  std::optional<std::tuple<uint64_t, uint64_t, uint64_t>> core::get_coinbase_tx_sum(uint64_t start_offset, size_t count)
   {
-    uint64_t emission_amount = 0;
-    uint64_t total_fee_amount = 0;
-    uint64_t burnt_loki = 0;
-    if (count)
-    {
-      const uint64_t end = start_offset + count - 1;
-      m_blockchain_storage.for_blocks_range(start_offset, end,
-        [this, &emission_amount, &total_fee_amount, &burnt_loki](uint64_t, const crypto::hash& hash, const block& b){
+    std::tuple<uint64_t, uint64_t, uint64_t> result{0, 0, 0};
+    if (count == 0)
+      return result;
+
+    auto& [emission_amount, total_fee_amount, burnt_loki] = result;
+
+    // Caching.
+    //
+    // Requesting this value from the beginning of the chain is very slow, so we cache it.  That
+    // still means the first request will be slow, but that's okay.  To prevent a bunch of threads
+    // getting backed up trying to calculate this, we lock out more than one thread building the
+    // cache at a time if we're requesting a large number of block values at once.  Any other thread
+    // requesting will get a nullopt back.
+
+    constexpr uint64_t CACHE_LAG = 30; // We cache the values up to this many blocks ago; we lag so that we don't have to worry about small reorgs
+    constexpr uint64_t CACHE_EXCLUSIVE = 1000; // If we need to load more than this, we block out other threads
+
+    // Check if we have a cacheable from-the-beginning result
+    uint64_t cache_to = 0;
+    std::chrono::steady_clock::time_point cache_build_started;
+    if (start_offset == 0) {
+      uint64_t height = m_blockchain_storage.get_current_blockchain_height();
+      if (count > height) count = height;
+      cache_to = height - std::min(CACHE_LAG, height);
+      {
+        std::shared_lock lock{m_coinbase_cache.mutex};
+        if (count >= m_coinbase_cache.height) {
+          emission_amount = m_coinbase_cache.emissions;
+          total_fee_amount = m_coinbase_cache.fees;
+          burnt_loki = m_coinbase_cache.burnt;
+          start_offset = m_coinbase_cache.height;
+          count -= m_coinbase_cache.height;
+        }
+        // else don't change anything; we need a subset of blocks that ends before the cache.
+
+        if (cache_to <= m_coinbase_cache.height)
+          cache_to = 0; // Cache doesn't need updating
+      }
+
+      // If we're loading a lot then acquire an exclusive lock, recheck our variables, and block out
+      // other threads until we're done.  (We don't do this if we're only loading a few because even
+      // if we have some competing cache updates they don't hurt anything).
+      if (cache_to > 0 && count > CACHE_EXCLUSIVE) {
```

Comment:
imo this adds a lot of complexity for little gain to handle another thread executing the same time here. I think it'd not be the end of the world to lock and snap a copy of the statistics if the cache is applicable and build the rest missing + the lock on `building` to fix that 1st time call not-niceness.

Is there a common case where this function is used typically? I think the explorer caches this value

Reply:
You never want two calls in here at once because *each* call is blocking an RPC thread on a long operation, so if you have something that tries to call this often we end up with a lokid that stays locked for a long time.  The use case here is being able to make a block explorer that doesn't need direct lmdb and internal lokid function access or need to have long running threads, and can hit this RPC call once to start calculating it, and later can start retrieving it once it becomes available.  I suppose the complexity here could move into the RPC code, but since that's the only code that ever calls this I'm not sure it makes much difference.

Reply:
Oh this part doesn't call unless `count > CACHE_EXCLUSIVE`. I thought this was doing a re-check on the cache even if only a couple of blocks changed so I was thinking it was a micro-optimisation.